ee/clickhouse/queries/diagnose/diagnose.py

```diff
@@ -11,4 +19,76 @@ def __init__(self, filter: DiagnoseFilter, team: Team) -> None:
         pass
 
     def run(self, *args, **kwargs) -> DiagnoseResponse:
-        return {"events": [{"event": "watch video", "value": 1}]}
+        """
+        Run the diagnose query.
+
+        Diagnose queries take as input:
+
+         - A persons filter criteria (specifying our cohort of interest)
+         - A success criteria (for instance a "signup" event)
+         - Dimensions along which to group people (for instance "browser" or
+           "os", or emitted event "played video")
+
+        Roughly speaking, to calculate the odds ratio, we build a contingency
+        table https://en.wikipedia.org/wiki/Contingency_table for each
+        dimension, then calculate the odds ratio for each.
+
+        For example, take for simplicity the cohort of all people, and the
+        success criteria of having a "signed up" event. First we would build a
+        contingency table like:
+
+        |                    | success | failure | total |
+        | -----------------: | :-----: | :-----: | :---: |
+        | watched video      |    5    |    1    |   6   |
+        | didn't watch video |    2    |   10    |   12  |
+
+
+        Then the odds that a person signs up given they watched the video is 5 /
+        6.
+        
+        And the odds that a person signs up given they didn't watch the video is
+        2 / 12.
+
+        So we say the odds ratio is 5 / 6 over 2 / 12 = 5 . The further away the
+        odds ratio is from 1, the greater the correlation.
+
+        Requirements:
+
+         - Intitially we only need to consider the names of events that a cohort
+           person has emitted. So we explicitly are not interested in e.g.
+           correlating properties, although this will be a follow-up.
+         - We only want to consider events that happened prior to the success
+           event, or the before the specified `date_to`, whichever comes first
+           on a per person basis
+
+        Non-functional requirements:
+
+         - there can be perhaps millions of people in a cohort, so we should
+           consider this when writing the algorithm. e.g. we should probably
+           avoid pulling all people into across the wire.
+         - there can be an order of magnitude more events than people, so we
+           should avoid pulling all events across the wire.
+         - there may be a large but not hug number of distinct events, let's say
+           100 different names for events. We should avoid n+1 queries for the
+           event names dimension
+
+        Contincency tables are something we can pull out of the db, so we can
```

Comment:
This is AMAZING. Thanks for writing this up! :D

And yes, this approach makes sense, with a caveat:

It's not just cohort criteria, but a funnel step, or a Paths step that can determine these people. Right now, we're only dealing with funnels, so makes sense to not-generalise, since we're not yet sure how things will look for the rest.

(Cohort and Funnel step have different semantics: A cohort is just a list of people, while a funnel step is a list of people + specific timestamps in their journey through the funnel - so you literally need the funnel query to get the people & their timestamps)


Also note that, we're not just looking for success, but failure odds too! (It can make sense to separate these out when dealing with too many event names - so you get the top of the list for success counts, and bottom of the list for failure counts, instead of getting the entire thing and trying to sort in Python). Worth adding to requirements.


```suggestion
        Contingency tables are something we can pull out of the db, so we can
```